src/python/pants/backend/docker/parser.py

```diff
@@ -0,0 +1,105 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import re
+from dataclasses import dataclass
+from typing import Generator
+
+from dockerfile import Command, parse_string
+
+from pants.backend.docker.target_types import DockerImageSources
+from pants.engine.fs import DigestContents, PathGlobs
+from pants.engine.rules import Get, collect_rules, rule
+from pants.option.global_options import GlobalOptions
+
+
+@dataclass(frozen=True)
+class DockerfileParseRequest:
+    sources: DockerImageSources
+
+
+@dataclass(frozen=True)
+class DockerfileInfo:
+    putative_target_addresses: tuple[str, ...] = ()
+
+
+# We pay for the up front time it takes to compile this at load time, as the alternative complicates
+# the implementation. See PR #12920.
```

Comment:
No need for this comment TBH, since we're doing the normal, obvious thing. I think comments explaining why the code is like it is and not like X only really make sense when X is the obvious thing...